### ISSUE TITLE
Reaction templates

### ISSUE THREAD

**Issue by mrsfy** [iVis-at-Bilkent/chise.js#8]
_From @ugurdogrusoz on April 22, 2016 12:33_

This high priority feature is to enable users to easily create certain types of processes such as associations and dissociations using templates provided. This is detailed out in Metin's GSoC project proposal.

_Copied from original issue: iVis-at-Bilkent/sbgnviz-js#118_

**Comment by mrsfy** [comment]
_From @ugurdogrusoz on June 1, 2016 6:40_

Let's start by designing a UI for the two templates we know of (association and dissociation). Let's discuss please.

**Comment by mrsfy** [comment]
_From @metincansiper on June 9, 2016 11:11_

I have enabled this feature for association and dissociation. You can see the related screenshots below.

Association Menu Screen
![association-menu](https://cloud.githubusercontent.com/assets/6534865/15927683/bc618d0c-2e4b-11e6-8fdc-b5b8559c5fe6.PNG)

Dissociation Menu Screen
![dissociation-menu](https://cloud.githubusercontent.com/assets/6534865/15927694/c1c60a70-2e4b-11e6-84e6-1a9cf399b919.PNG)

Association Example
![association](https://cloud.githubusercontent.com/assets/6534865/15927697/c6e84388-2e4b-11e6-9369-edc9b3fd3fa1.PNG)

Dissociation Example
![dissociation](https://cloud.githubusercontent.com/assets/6534865/15927703/cb85ba60-2e4b-11e6-9994-43e049eb59d3.PNG)

**Comment by mrsfy** [comment]
_From @ugurdogrusoz on June 15, 2016 7:59_

@ozgunbabur Could you please check if the implementation is OK or can be improved in any way?
Edit > Add From Template...

**Comment by mrsfy** [comment]
_From @ozgunbabur on June 17, 2016 18:32_

It looks nice and simple, with below concerns:

1. What if one of the members is a complex? For instance A-B + C --> A-B-C. Are you planning to support complexes as members?

2. The checkbox should say something like "Name the complex".

3. The resulting graph has fixed node dimensions, making node names unreadable.

![image](https://cloud.githubusercontent.com/assets/1376779/16160600/f8141224-347d-11e6-9a4e-37549b69a021.png)

4. The generated association or dissociation node size is not compatible with what we already have on the graph. Highlighted portion below is created with association template.

![image](https://cloud.githubusercontent.com/assets/1376779/16160711/6ac63568-347e-11e6-9c69-8814a2483209.png)

5. Most of the times, we want to use existing molecules in these association and dissociations. There should be some mechanism to support using existing nodes, or alternatively, post processing by a feature that lets user to "merge" two molecules indicating they are the same thing.

**Comment by ugurdogrusoz** [comment]
Outstanding issues:
- You need to click Create button twice!
- Icons for delete don't have proper aspect ratio
- Column names (Input, Output) need some space before them
Let's not worry about using existing EPNs for now.
Let's discuss if there is any other template that would be useful.

**Comment by royludo** [comment]
Above mentioned issues are fixed in  iVis-at-Bilkent/newt@9cf36baee22109e66cb948626d84aed8b2160923
I completely rewrote this dialog as it was a bit weirdly coded before, which was leading to the clicking-twice issue. Comparing before and now:
 - I got rid of the checkbox to enable complex name, as it's something we can directly check (if text box is empty, no complex name). Was this checkbox really needed ? I think we can put a placeholder/watermark/grey text in the text box to indicate that it is for complex name.
 - now the dialog never remembers user inputs. It always gets reset when it disappear. Before, inputs were remembered until the user hit the Create button, and then it was reset. I didn't feel like remembering inputs here was really needed, but I can put it back if we decide it is.

Regarding other templates, Alexander told me about a [publication](http://www.sciencedirect.com/science/article/pii/S0167779912001333) he did that was listing what he called SBGN bricks. There's a wiki page [on sourceforge](http://sbgnbricks.sourceforge.net/), with a link to download a zip file at the bottom. It contains sbgn maps of those bricks. They were implemented in SBGN-ED. But the thing is, we didn't really think it was something urgent. From our perspective, implementing more templates is something that can wait.
